whylog/constraints/__init__.py

```diff
@@ -2,42 +2,184 @@
 
 import six
 
+from whylog.constraints.const import ConstraintType
+from whylog.constraints.exceptions import ConstructorGroupsError, ConstructorParamsError
+from whylog.teacher.user_intent import UserConstraintIntent
+
 
 @six.add_metaclass(ABCMeta)
 class AbstractConstraint(object):
+    """
+    :param MIN_GROUPS_COUNT: minimal groups count needed to create constraint
+    :param MAX_GROUPS_COUNT: maximal groups count needed to create constraint
+    """
+
+    TYPE = 'undefined'
+    MIN_GROUPS_COUNT = None
+    MAX_GROUPS_COUNT = None
+
     @abstractmethod
-    def save(self, rulebase_rule):
-        pass
+    def __init__(self, param_dict, groups):
+        """
+        Creates constraint object.
+
+        For Teacher and Front use while creating user rule.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: all groups that are linked by constraint,
+                       represented by list of tuples (line_id, group_no),
+                       where line_id and group_no is inner numeration between Front and Teacher
+        """
+        self.params = param_dict
+        self.groups = groups
+
+    @abstractmethod
+    def convert_to_user_constraint_intent(self):
+        """
+        Converts constraint to UserConstraintIntent object.
+
+        For Teacher and Config use while saving constraint into Whylog knowledge base.
+        """
+        return UserConstraintIntent(self.TYPE, self.groups, self.params)
+
+    @classmethod
+    def get_param_names(cls):
+        """
+        Returns names of constraint additional params.
+
+        For Front to display param names to user and then ask user for param contents.
+        """
+        raise NotImplementedError("Subclass should implement this")
+
+    @classmethod
+    def get_groups_count(cls):
+        """
+        Returns minimal and maximal count of groups needed to create constraint.
+
+        For Front to ask user for proper count of groups.
+        2, None - at least 2 groups
+        2, 2 - exactly 2 groups
+        """
+        return cls.MIN_GROUPS_COUNT, cls.MAX_GROUPS_COUNT
+
+    @classmethod
+    def verify(cls, param_dict, group_contents):
+        """
+        Verifies constraint for given params in param_dict and groups contents.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: list of groups contents,
+
+        For LogReader and Teacher verification.
+        It must be optimized as well as possible (for LogReader).
+        """
+        raise NotImplementedError("Subclass should implement this")
```

Comment:
I agree that we should raise exception here. But I'm not sure if `NotImplementedError` is the best choice. But I saw in other fragments of code that in similar situation `NotImplementedError` is raised. So you can keep it.